# Data pipeline

In this notebook the data is collected and preprocessed before exploratory analysis can be performed.

The data is provided from the online Data Science platform kaggle.

## Loading in relevant packages

In [5]:
import pandas as pd
import numpy as np
import kagglehub
import os

## Fetching the data from kaggle.com

In [6]:
# Download latest version
path = kagglehub.dataset_download("meruvakodandasuraj/e-commerce-customer-behavior-and-sales-20202026")

print("Path to dataset files:", path)

Path to dataset files: C:\Users\niels\.cache\kagglehub\datasets\meruvakodandasuraj\e-commerce-customer-behavior-and-sales-20202026\versions\1


I find the files in the path

In [9]:
files = os.listdir(path)

Then the csv-files are identified

In [24]:
path

'C:\\Users\\niels\\.cache\\kagglehub\\datasets\\meruvakodandasuraj\\e-commerce-customer-behavior-and-sales-20202026\\versions\\1'

In [35]:
datasets = {csv[:-4]: pd.read_csv(f'{path}\\{csv}', sep = ",") for csv in files if '.csv' in csv}

## Inspection of each dataset

In [36]:
for key in datasets.keys():
    print('The dataset ' + key + ' contains the following columns:')
    for col in datasets[key].columns:
        print(col)
    print('\n')

The dataset customers contains the following columns:
customer_id
country
age
gender
membership_tier
registration_date
total_orders
total_spend_usd
avg_order_value_usd
days_since_last_purchase
preferred_category
preferred_device
preferred_payment_method
acquisition_channel
reviews_given
avg_review_score
returns_made
wishlist_items
newsletter_subscribed
churned


The dataset monthly_revenue contains the following columns:
year
month
quarter
orders
revenue_usd
avg_order_value
avg_discount_pct
return_rate
unique_customers
new_customers


The dataset orders contains the following columns:
order_id
customer_id
order_date
year
month
quarter
day_of_week
product_name
category
unit_price_usd
quantity
subtotal_usd
discount_pct
discount_amount_usd
shipping_fee_usd
tax_pct
tax_amount_usd
total_amount_usd
payment_method
device_used
delivery_days
delivery_date
order_status
returned
customer_rating
session_duration_minutes
pages_viewed_before_purchase
is_repeat_customer


The dataset product_summary 

We consider the above datasets to be our complete information regarding our loosely defined objective of modelling churn and additionally other relevant analysis.

In [42]:
for key in datasets.keys():
    print(f"The dataset {key} contains {sum(pd.isna(datasets[key]).sum())} missing values")

The dataset customers contains 0 missing values
The dataset monthly_revenue contains 0 missing values
The dataset orders contains 15749 missing values
The dataset product_summary contains 0 missing values


It turns out that 3 out of 4 datasets contains no missing values. The dataset orders contains 15,749 missing values spread across the following columns

In [43]:
orders = datasets["orders"]

print("Number of missings:")
for col in orders.columns:
    print(f"{col}: {sum(pd.isna(orders[col]))}")

Number of missings:
order_id: 0
customer_id: 0
order_date: 0
year: 0
month: 0
quarter: 0
day_of_week: 0
product_name: 0
category: 0
unit_price_usd: 0
quantity: 0
subtotal_usd: 0
discount_pct: 0
discount_amount_usd: 0
shipping_fee_usd: 0
tax_pct: 0
tax_amount_usd: 0
total_amount_usd: 0
payment_method: 0
device_used: 0
delivery_days: 0
delivery_date: 0
order_status: 0
returned: 0
customer_rating: 15749
session_duration_minutes: 0
pages_viewed_before_purchase: 0
is_repeat_customer: 0


The only column in orders containing missing is the column customer_ratings.

In [53]:
print(orders["customer_rating"].describe())

count    9251.000000
mean        4.002324
std         0.577576
min         1.500000
25%         3.600000
50%         4.100000
75%         4.400000
max         5.000000
Name: customer_rating, dtype: float64


In [57]:
print('Proportion of missings: ' + str(round(sum(pd.isna(orders["customer_rating"]))/len(orders["customer_rating"])*100,2)) + '%')

Proportion of missings: 63.0%


From the values and the column name it is likely a review going from between 1.5 to 5. It is yet unclear if it is the order, the customer experience, the product or the firm as a whole that is being reviewed, however it is likely a full customer experienced related to the order. 

Missings likely reflect customers who did not give a review of the order or an attempted review got lost in the data. 

This feature is hard imputate a substitute value as the average is a 4 out of 5 and might bias because of customers who had an excellent or a horrible experience are more likely to leave a review.  

For now the data set customers contains naive features we might consider while doing churn modelling hence we look at it first.

In [62]:
datasets["customers"].to_csv(r"C:\Users\niels\OneDrive\Dokumenter\Churn modelling\Churn-modelling\data\customers.csv", sep = ",", decimal=".")